pkg/cloudprovider/providers/gce/gce.go

```diff
@@ -724,6 +746,30 @@ func (gce *GCECloud) updateNodeZones(prevNode, newNode *v1.Node) {
 	}
 }
 
+func (gce *GCECloud) addNodeToInstanceMap(newNode *v1.Node) error {
+	gce.instanceNameMapLock.Lock()
+	defer gce.instanceNameMapLock.Unlock()
+
+	if newNode != nil {
+		_, _, instanceName, err := splitProviderID(newNode.Spec.ProviderID)
```

Comment:
this is changing the cloud controllers to *pull* information from the ProviderID field of a Node API object, correct? doesn't that make the kubelet the authority on which cloud provider instance a given Node API object corresponds to?
@mikedanese ^

Reply:
The only thing consumed from ProviderID field here is the instance name, which was assumed to be the node name before this by the cloud controllers. So, the kubelet was and still is the the authority in defining the node name, which was assumed to be the instance name by the cloud controllers.

Reply:
The kubelet does not get to choose any node metadata.name it wants. It is only allowed to modify the one that corresponds to its kube API credentials. 

If this is changing cloudprovider code to be driven by the provider id field instead, we first have to ensure a compromised kubelet cannot claim to be another instance (by setting the providerID field) and get the cloud controller (or other controllers) to take actions against that instance

Reply:
Hmm, I get it now. I was always assuming that kubelets were trusted as far as they were authenticated. I am not sure how to solve this problem, I will try to dig into other cloud providers to see how they solve it. Meanwhile, if you have any suggestions, please let me know.

Reply:
I checked in other cloud providers (AWS, Openstack, Azure and vSphere), it seems like only vSphere relies on ProviderID and hence is insecure against a node being compromised. 

After chatting with a few people an idea that seems to work is tagging nodes in IAAS to make sure kubelet is not telling cloud controllers which nodes it corresponds to. For example, `node-1` runs on instance `vm-foo`. In this case `vm-foo` will be expected to have a tag say `kubernetes-node-name: node-1`. When node-1 joins the cluster, the cloud provider will look for vm with the said tag and find the instance name, from this point on the cloud provider can keep a mapping of node name to instance name.

This seems to solve the trust issue between master components and kubelet, assuming the tags can be trusted. But it will require operator to make extra configurations. I am not sure if that is acceptable. WDYT?

Reply:
With labels, the code for `addNodeToInstanceMap` will look like this:
```golang
	if newNode != nil {
		_, zone, instanceName, err := splitProviderID(newNode.Spec.ProviderID)
		if err != nil {
			return err
		}
		instance, err := gce.Compute().Instances().Get(context.TODO(), &meta.Key{Name: instanceName, Zone: zone})
		if err != nil {
			return err
		}
		allowedNodeName, ok := instance.Labels["kubernetes-node-name"]
		if !ok {
			return errors.Errorf("Could not find label: 'kubernetes-node-name' on instance '%s'", instanceName)
		}

		if allowedNodeName != newNode.Name {
			return errors.Errorf("Instance '%s' is only allowed to have node name '%s'", instanceName, allowedNodeName)
		}

		gce.instanceNameMap[allowedNodeName] = instanceName
	}
```

Reply:
> an idea that seems to work is labelling nodes in IAAS to make sure kubelet is not telling cloud controllers which nodes it corresponds to

having data living in the cloud provider being the source of truth is much closer to what I expected.

with that information living there, is it actually still necessary to have the ProviderID in the Node object? it's tough to ensure that every component that makes use of that field does a verification against a central source of truth that the mapping is actually correct.

Reply:
I can think of two reasons for ProviderID to be on the Node Object:
1. It helps the cloud provider look at just one node and verify if the information is correct.
1. If somebody wants to recreate the worker while previous one is not gone, they can smoothly switch over by creating the new worker and then killing the previous one.

>it's tough to ensure that every component that makes use of that field does a verification against a central source of truth that the mapping is actually correct

No component (except for the `addNodeToInstanceMap` method) should be using ProviderID directly, they should always use the `mapNodeNameToInstanceName` method to get the instance name and the method above will make sure that the mapping is always verified.

Reply:
I agree, it is weird to have that implicit assumption. Maybe the two points I listed above could be handled in other ways. 

It seems to be a little tricky to list VMs, because the GCE API to list vms only lists VMs in a given zone. So, either we'll need to make the `Zones` field mandatory in the cloud config or look for the instance through all the zones of GCE, which sounds like an overkill.